File: backend/jwt/jwt_service.go
package jwt_service

import (
	"os"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

/*
Файл jwt_service.go содержит функции для генерации JWT-токенов
проекта. Основной токен авторизации (30 дней) и отдельный токен
для процедуры сброса пароля (15 минут). Парсинг и чтение токенов
происходит в services, здесь только их создание и подпись
*/

/*
GenerateJWT создаёт основной JWT-токен авторизации
с полями user_id и exp (30 дней). Минимум данных в токене,
так как все остальное достается из БД через /api/me
*/
func GenerateJWT(userID int) (string, error) {
	claims := jwt.MapClaims{
		"user_id": userID,
		"exp":     time.Now().Add(30 * 24 * time.Hour).Unix(),
	}
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString([]byte(os.Getenv("JWT_SECRET")))
}

/*
GenerateResetPasswordJWT создаёт отдельный JWT-токен
для процедуры сброса пароля с полями email и purpose
(15 минут). Подписан тем же JWT_SECRET, что и основной
токен, поэтому email внутри нельзя подделать, а purpose
не даёт использовать обычный токен авторизации вместо
токена сброса пароля
*/
func GenerateResetPasswordJWT(email string) (string, error) {
	claims := jwt.MapClaims{
		"email":   email,
		"purpose": "reset_password",
		"exp":     time.Now().Add(15 * time.Minute).Unix(),
	}
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString([]byte(os.Getenv("JWT_SECRET")))
}

File: backend/jwt/jwt_service.rs
use chrono::Utc;
use jsonwebtoken::{decode, encode, DecodingKey, EncodingKey, Header, Validation};
use serde::{Deserialize, Serialize};

use crate::config;

#[derive(Debug, Serialize, Deserialize)]
pub struct AuthClaims {
    pub user_id: i32,
    pub exp: usize,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ResetPasswordClaims {
    pub email: String,
    pub purpose: String,
    pub exp: usize,
}

pub fn generate_jwt(user_id: i32) -> Result<String, jsonwebtoken::errors::Error> {
    let exp = (Utc::now() + chrono::Duration::days(30)).timestamp() as usize;
    let claims = AuthClaims { user_id, exp };
    encode(
        &Header::default(),
        &claims,
        &EncodingKey::from_secret(config::get_jwt_secret().as_bytes()),
    )
}

pub fn generate_reset_password_jwt(email: &str) -> Result<String, jsonwebtoken::errors::Error> {
    let exp = (Utc::now() + chrono::Duration::minutes(15)).timestamp() as usize;
    let claims = ResetPasswordClaims {
        email: email.to_string(),
        purpose: "reset_password".to_string(),
        exp,
    };
    encode(
        &Header::default(),
        &claims,
        &EncodingKey::from_secret(config::get_jwt_secret().as_bytes()),
    )
}

pub fn decode_auth_jwt(token: &str) -> Result<AuthClaims, jsonwebtoken::errors::Error> {
    let data = decode::<AuthClaims>(
        token,
        &DecodingKey::from_secret(config::get_jwt_secret().as_bytes()),
        &Validation::default(),
    )?;
    Ok(data.claims)
}

pub fn decode_reset_password_jwt(
    token: &str,
) -> Result<ResetPasswordClaims, jsonwebtoken::errors::Error> {
    let data = decode::<ResetPasswordClaims>(
        token,
        &DecodingKey::from_secret(config::get_jwt_secret().as_bytes()),
        &Validation::default(),
    )?;
    Ok(data.claims)
}
